module/zstd/zfs_zstd.c

```diff
@@ -688,6 +690,19 @@ zstd_mempool_deinit(void)
 	zstd_mempool_cctx = NULL;
 }
 
+static boolean_t
+zstd_memory_liberator_cb(void *arg, zthr_t *zthr)
+{
+	/*
+	 * calling alloc with zero size seeks
+	 * and releases orphan objects
+	 */
+	zstd_mempool_alloc(zstd_mempool_cctx, 0);
+	zstd_mempool_alloc(zstd_mempool_dctx, 0);
+	return (B_FALSE);
+}
```

Comment:
Instead of moving the zthr code and creating another thread how about leveraging the existing `arc_reap` thread which already performs this same function.  This thread calls the `arc_kmem_reap_soon()` function which reaps unused objects from a variety of ZFS related caches.  You'd just need to add a `zstd_cache_reap_now()` function and call it.

Reply:
@behlendorf circular dependency. you cannot load zfs without zsstd beeing loaded. this is the same reason why i moved zthr.c. so no call to zzstd can be made from code within zfs

Reply:
I'm not sure I follow.   I was suggesting that `arc_kmem_reap_soon()` call down in to the `zzstd` module via a new public zstd function, say `zfs_zstd_cache_reap_now()`.

Reply:
ahm sorry. yes. now i understand. my fault

Reply:
i have one problem with arc_kmem_reap_soon etc. the rule how arc_cb_reap is called is bad, it exclusive checks rules related to the available arc memory which has nothing todo with a cyclic cleanup procedure. it triggers only when memory is out and under pressure. consider that my release function releases objects which are aged already. so a pure cylic call is all i need. but  arc_kmem_reap_soon will never be called if the system still has memory left

Reply:
Right, I agree it's a bit awkward.  However, I think even calling `zfs_zstd_cache_reap_now()` every second would be pretty reasonable, particularly if the size of the pool were known and we could return immediately when it was empty.  Adding this same size information to the zstd_stats_t kstat would also be really helpful to make easy to check how much memory is really being used here.

Reply:
@behlendorf consider that i iterate the pool which is small of course, but i have to call a trylock for each object. so latency is everything here and since aged/outdated mempool objects is defined with "unused since more than 2 minutes) checking every second is not required. of course it would react fast, but it doesnt need so. with normal workflow this polling is also not required since checking for orphan objects is done on each allocation attempt. this cleanup procedure is only required if all io access has been stopped or finished related to zstd compression. so is there any hurry? surelly not.

Reply:
> this cleanup procedure is only required if all io access has been stopped or finished related to zstd compression

It would potentially be desirable to be able to reclaim this space when free memory is low, but changing the polling frequency of course wouldn't help since it won't age out for 2 minutes.  Let's go ahead and put this reclaim mechanism is place so it will get reclaimed and we can always refine it.

Reply:
regarding the way when to reclaim space and when not. the pool allocator is self managing. waiting until memory is low, if avoidable is not a good way from my point of view. in our testcase we had about 1.5 gb unused memory left. thats no small piece. releasing it if not in use for a while avoids such situations which will slow down zfs. so if avoidable without performance impact is a good approach